ledger/ledger.go

```diff
@@ -183,15 +184,68 @@ func (k *Key) Size() int {
 	return size
 }
 
+// this is a helper function to optimize CanonicalForm
+// this will store the Type and Value in the given byte slice,
+// using the format '/1234/value'
+// It returns the exact size of the string
+func (kp *KeyPart) store(arr []byte, offset int) int {
+	typeNumber := strconv.Itoa(int(kp.Type))
+	n := offset
+
+	arr[n] = byte('/')
+	n++
+	lenType := len(typeNumber)
+	lenValue := len(kp.Value)
+
+	// copy the number as text into the slice
+	for i := 0; i < lenType; i++ {
+		arr[n] = typeNumber[i]
+		n++
+	}
+
+	// copy the value into the slice
+	arr[n] = byte('/')
+	n++
+	for i := 0; i < lenValue; i++ {
+		arr[n] = kp.Value[i]
+		n++
+	}
+
+	// return the actual size
+	return n - offset
+}
+
 // CanonicalForm returns a byte slice describing the key
 // Warning, Changing this has an impact on how leaf hashes are computed
 // don't use this to reconstruct the key later
 func (k *Key) CanonicalForm() []byte {
-	ret := ""
+	// calculate the size of the byte array
+
+	// the maximum size of a uint16 is 5 characters, so
+	// this is using 10 for the estimate, to include the two '/'
+	// characters and an extra 3 characters for padding safety
+
+	requiredLen := 0
+	for _, kp := range k.KeyParts {
+		requiredLen += 10
+		requiredLen += len(kp.Value)
+	}
+	retval := make([]byte, requiredLen)
+
+	// get the exact size, while writing the key parts
+	requiredLen = 0
 	for _, kp := range k.KeyParts {
-		ret += fmt.Sprintf("/%d/%v", kp.Type, string(kp.Value))
+		requiredLen += kp.store(retval, requiredLen)
+	}
+
+	// create a byte slice with the correct size and copy
+	// the estimated string into it.
+	corrected := make([]byte, requiredLen)
+	for i := 0; i < requiredLen; i++ {
+		corrected[i] = retval[i]
```

Comment:
- we could do :  `corrected := retval[:requiredLen]` if we don't care about reusing the pointer.
- If we really care about the pointer, we should do `copy(corrected, retval)` and avoid the non-optimized copy by byte.
- and after reading all the function, I am wondering why don't we create one slice only `retval` and make the `store` function append to it. This avoids allocating 2 slices, avoids the extra copy, the estimation computation of `requiredLen`..

Reply:
I tried with append(), but it was slower than this version.